File: apps/api/internal/http/dto/eta_reliability.go
package dto

import (
	"time"

	"github.com/AsifAbbasov/global-flight-analytics/apps/api/internal/projectionintelligence/etareliability"
)

type ETAReliabilityRoute struct {
	OriginICAOCode string `json:"origin_icao_code"`
	DestinationICAOCode string `json:"destination_icao_code"`
}

type ETAReliabilityMethod struct {
	Name string `json:"name"`
	Version string `json:"version"`
	DecisionClass string `json:"decision_class"`
}

type ETAReliabilityMetrics struct {
	SampleCount int `json:"sample_count"`
	MedianAbsoluteErrorSeconds float64 `json:"median_absolute_error_seconds"`
	P80AbsoluteErrorSeconds float64 `json:"p80_absolute_error_seconds"`
	WithinFiveMinutesRatio float64 `json:"within_five_minutes_ratio"`
	WithinTenMinutesRatio float64 `json:"within_ten_minutes_ratio"`
	IntervalCoverageRatio float64 `json:"interval_coverage_ratio"`
}

type ETAReliabilityNotice struct {
	Code string `json:"code"`
	Message string `json:"message"`
}

type ETAReliabilityResponse struct {
	Version string `json:"version"`
	Status string `json:"status"`
	TrajectoryID string `json:"trajectory_id"`
	Route ETAReliabilityRoute `json:"route"`
	Method ETAReliabilityMethod `json:"method"`
	TargetLeadSeconds int64 `json:"target_lead_seconds"`
	LeadToleranceSeconds int64 `json:"lead_tolerance_seconds"`
	EndpointRadiusKM float64 `json:"endpoint_radius_km"`
	CandidateCount int `json:"candidate_count"`
	EligibleSampleCount int `json:"eligible_sample_count"`
	Metrics *ETAReliabilityMetrics `json:"metrics,omitempty"`
	EvidenceClass string `json:"evidence_class"`
	Limitations []ETAReliabilityNotice `json:"limitations"`
	InputFingerprint string `json:"input_fingerprint"`
	GeneratedAt time.Time `json:"generated_at"`
}

func ToETAReliabilityResponse(result etareliability.Result) ETAReliabilityResponse {
	response := ETAReliabilityResponse{
		Version: result.Version,
		Status: string(result.Status),
		TrajectoryID: result.TrajectoryID,
		Route: ETAReliabilityRoute{
			OriginICAOCode: result.Route.OriginICAOCode,
			DestinationICAOCode: result.Route.DestinationICAOCode,
		},
		Method: ETAReliabilityMethod{
			Name: result.Method.Name,
			Version: result.Method.Version,
			DecisionClass: string(result.Method.DecisionClass),
		},
		TargetLeadSeconds: result.TargetLeadSeconds,
		LeadToleranceSeconds: result.LeadToleranceSeconds,
		EndpointRadiusKM: result.EndpointRadiusKM,
		CandidateCount: result.CandidateCount,
		EligibleSampleCount: result.EligibleSampleCount,
		EvidenceClass: result.EvidenceClass,
		InputFingerprint: result.InputFingerprint,
		GeneratedAt: result.GeneratedAt.UTC(),
		Limitations: make([]ETAReliabilityNotice, 0, len(result.Limitations)),
	}
	for _, item := range result.Limitations {
		response.Limitations = append(response.Limitations, ETAReliabilityNotice{Code: item.Code, Message: item.Message})
	}
	if result.Metrics != nil {
		response.Metrics = &ETAReliabilityMetrics{
			SampleCount: result.Metrics.SampleCount,
			MedianAbsoluteErrorSeconds: result.Metrics.MedianAbsoluteErrorSeconds,
			P80AbsoluteErrorSeconds: result.Metrics.P80AbsoluteErrorSeconds,
			WithinFiveMinutesRatio: result.Metrics.WithinFiveMinutesRatio,
			WithinTenMinutesRatio: result.Metrics.WithinTenMinutesRatio,
			IntervalCoverageRatio: result.Metrics.IntervalCoverageRatio,
		}
	}
	return response
}

File: apps/web/types/eta-reliability.ts
export type ETAReliabilityStatus = 'unavailable' | 'limited' | 'complete'

export interface ETAReliabilityRoute {
  origin_icao_code: string
  destination_icao_code: string
}

export interface ETAReliabilityMethod {
  name: string
  version: string
  decision_class: string
}

export interface ETAReliabilityMetrics {
  sample_count: number
  median_absolute_error_seconds: number
  p80_absolute_error_seconds: number
  within_five_minutes_ratio: number
  within_ten_minutes_ratio: number
  interval_coverage_ratio: number
}

export interface ETAReliabilityNotice {
  code: string
  message: string
}

export interface ETAReliabilityResponse {
  version: string
  status: ETAReliabilityStatus
  trajectory_id: string
  route: ETAReliabilityRoute
  method: ETAReliabilityMethod
  target_lead_seconds: number
  lead_tolerance_seconds: number
  endpoint_radius_km: number
  candidate_count: number
  eligible_sample_count: number
  metrics?: ETAReliabilityMetrics
  evidence_class: 'historically_recomputed_from_persisted_observations_with_endpoint_proxy'
  limitations: ETAReliabilityNotice[]
  input_fingerprint: string
  generated_at: string
}

export interface ETAReliabilityRequest {
  trajectoryID: string
  asOfTime: string
  durationSeconds: number
  signal?: AbortSignal
}
